crates/ruff/src/jupyter/notebook.rs

```diff
@@ -130,60 +173,201 @@ impl JupyterNotebook {
             )));
         }
 
-        Ok(notebook)
-    }
-
-    /// Concatenates all cells into a single virtual file and builds an index that maps the content
-    /// to notebook cell locations
-    pub fn index(&self) -> (String, JupyterIndex) {
-        let mut jupyter_index = JupyterIndex {
-            // Enter a line number (1-based), get back the cell (1-based)
-            // 0 index is just padding
-            row_to_cell: vec![0],
-            // Enter a line number (1-based), get back the row number in the cell (1-based)
-            // 0 index is just padding
-            row_to_row_in_cell: vec![0],
-        };
-        let size_hint = self
+        let valid_code_cells = notebook
             .cells
             .iter()
-            .filter(|cell| cell.cell_type == CellType::Code)
-            .count();
+            .enumerate()
+            .filter(|(_, cell)| cell.is_valid_code_cell())
+            .map(|(pos, _)| u32::try_from(pos).unwrap())
+            .collect::<Vec<_>>();
 
-        let mut contents = Vec::with_capacity(size_hint);
+        let mut contents = Vec::with_capacity(valid_code_cells.len());
+        let mut current_offset = TextSize::from(0);
+        let mut cell_offsets = Vec::with_capacity(notebook.cells.len());
+        cell_offsets.push(TextSize::from(0));
 
-        for (pos, cell) in self
-            .cells
+        for &pos in &valid_code_cells {
+            let cell_contents = match &notebook.cells[pos as usize].source {
+                SourceValue::String(string) => string.clone(),
+                SourceValue::StringArray(string_array) => string_array.join(""),
+            };
+            current_offset += TextSize::of(&cell_contents) + TextSize::new(1);
+            cell_offsets.push(current_offset);
+            contents.push(cell_contents);
```

Comment:
Nit: maybe move this up a line? To match the logic: "Push the contents to the current cell, and push the starting offset of the _next_ cell."